### ISSUE TITLE
[Bug] Paged KV allocator launches allocation kernels before checking OOM

### ISSUE THREAD

**Issue by waizuichougou** [sgl-project/sglang#34399]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version (`a23670ddbf`, verified on 2026-08-13).
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

On the paged CUDA KV-cache path, `PagedTokenToKVPoolAllocator.alloc_extend`
and `alloc_decode` launch Triton allocation kernels before checking whether
`free_pages` contains enough pages.

The kernels index `free_pages` using computed page offsets, but the accesses
are not masked by the logical free-list length. When the KV cache is exhausted,
the kernel can read beyond the valid logical free-page list before the caller
returns the expected out-of-memory error.

The visible result is normally a controlled scheduler OOM, but the preceding
GPU read is invalid with respect to the allocator's logical free-page list and
may cause undefined behavior or poison the CUDA worker. This report does not
claim that every allocator layout produces a deterministic CUDA
illegal-address exception.

### Reproduction

The latest upstream `main` at commit
`a23670ddbf89678a53230ceb2cc5ed75236c4d85` still reproduces the issue.

#### Minimal CPU reproduction

With `page_size=4` and an allocator containing two free pages, the request
below requires three new pages. The kernel is replaced with a mock, so this
checks the Python-to-kernel ordering without requiring a model:

```python
from unittest.mock import patch

import torch

from sglang.srt.mem_cache.allocator.paged import PagedTokenToKVPoolAllocator

allocator = PagedTokenToKVPoolAllocator(
    size=8,
    page_size=4,
    dtype=torch.float16,
    device="cpu",
    kvcache=None,
    need_sort=False,
)

prefix_lens = torch.tensor([0], dtype=torch.int64)
seq_lens = torch.tensor([12], dtype=torch.int64)
last_loc = torch.tensor([-1], dtype=torch.int64)

with patch("sglang.srt.mem_cache.allocator.paged.alloc_extend_kernel") as kernel:
    result = allocator.alloc_extend(
        prefix_lens,
        prefix_lens,
        seq_lens,
        seq_lens,
        last_loc,
        extend_num_tokens=12,
    )

print(result is None, kernel.__getitem__.call_count)
```

On the latest unmodified `main` this prints:

```text
True 1
```

The expected result after the fix is:

```text
True 0
```

The allocator should return OOM before indexing the kernel launcher.

#### A100 reproduction

After warming the successful path, I wrapped the Triton kernel launchers and
ran the failing calls with `CUDA_LAUNCH_BLOCKING=1`.

Latest upstream `main`:

```text
gpu= NVIDIA A100-PCIE-40GB torch_cuda= 12.8
main_sha=a23670ddbf
extend_oom, kernel_launches= (True, 1)
decode_oom, kernel_launches= (True, 1)
```

The rebased branch in [PR #34400](https://github.com/sgl-project/sglang/pull/34400):

```text
gpu= NVIDIA A100-PCIE-40GB torch_cuda= 12.8
patch_sha=1af8b575ed
extend_oom, kernel_launches= (True, 0)
decode_oom, kernel_launches= (True, 0)
```

The same A100 inputs therefore return the same controlled OOM result without
launching either allocation kernel after the fix.

The focused regression tests on the patched branch pass:

```text
6 passed, 1 warning in 4.42s
```

#### Logical free-list over-read evidence

Using a backing tensor `[7, 99, 101]` and exposing only `backing[:1]` as the
logical `free_pages` tensor produced:

```text
logical_free_pages= [7]
page_ids_used= [7, 99, 101]
```

The backing tensor keeps this reproducer deterministic and in-bounds for the
underlying storage. It demonstrates that the kernel reads values beyond the
logical free-page list without claiming that every allocator layout produces
the same physical memory fault.

#### Online relevance

This condition is reachable during normal serving when the KV cache is full
and no more evictable pages are available.

The normal path is:

1. The scheduler calls `alloc_paged_token_slots_extend` or
   `alloc_paged_token_slots_decode`.
2. The allocator launches the Triton kernel while the free-page list is too
   short.
3. The allocator returns `None`, and the scheduler raises the expected OOM
   error.

Existing serving reports show the same allocator OOM paths:

- [#13765](https://github.com/sgl-project/sglang/issues/13765) reports
  `alloc_paged_token_slots_extend` failing during EAGLE verify with
  `available_size=448` and `evictable_size=0`.
- [#11581](https://github.com/sgl-project/sglang/issues/11581) reports Decode
  OOM in a running H100 service with `available_size=3` and
  `evictable_size=0`.

### Expected behavior

The allocator should check page capacity before allocating the output tensor or
launching the Triton kernel.

When capacity is insufficient, it should return `None` without consuming pages
or launching the kernel. With `need_sort=True`, the existing release-page merge
may still run before this check.

### Proposed fix

Move the existing `num_new_pages > len(self.free_pages)` checks before the
Triton kernel launches in both `alloc_extend` and `alloc_decode`.

### Environment

- SGLang main commit: `a23670ddbf89678a53230ceb2cc5ed75236c4d85`
- Python: 3.10
- PyTorch: 2.9.1+cu128
- CUDA runtime: 12.8
- GPU: NVIDIA A100-PCIE-40GB
- Compute capability: 8.0
- Triton: 3.5.1
- sglang-kernel: 0.4.6.post1

**Comment by waizuichougou** [comment]
A corresponding implementation PR is open at #34400. It moves the existing paged-capacity checks before both Triton allocation kernel launches and adds CPU regression tests plus A100 validation.
